docs/enhancements/fep-000-api-device-fleet.md

```diff
@@ -0,0 +1,487 @@
+certificate identification, an user can access any device rendered spec for debug purposes.
+
+The agent can pass: `resourceVersion` if this opaque field has not changed the API will return
+204 No Content, if the field has changed, the API will return the rendered spec.
+
+Change proposed:
+* The parameter name `fleet` is changed to `owner` to be consistent with the metadata field.
+
+Behavior changes:
+* If a device configuration is not yet rendered for some reason, FC will queue the device for rendering and instruct
+  the device to try again later by returning 202 Accepted.
+
+The user can find the fully rendered device spec by using the following cmdline command:
+```shell
+flightctl get device/xxxxxxx --rendered
+```
+
+or
+
+```shell
+flightctl get device/xxxxxxx --rendered -o json
+```
+
+#### `/device/{name}/status` (device side)
+
+This endpoint is used by the device agent to report the status of the device, this endpoint is write-only,
+each device can only access it's own device status based on the device certificate identification.
+
+The result of this write will be exposed on the `/device/{name}` endpoint as part of the status field.
+
+Reporting to this endpoint also serves as a heartbeat mechanism, the device agent should report the status,
+if a device stops reporting to this endpoint in the expected timelines it will be considered offline.
+
+### Fleet API
+
+Fleet is an automation layer on top of Device API, the DeviceTemplate is a mechanism to control drift
+and enforce policy across devices in a fleet.
+
+A DeviceTemplate is a template for a device spec, it's a way to define group specific configuration.
+
+We are proposing extension mechanisms to this API to allow for more complex configurations where
+device labels or name would influence the configuration generated to the device.
+
+https://github.com/flightctl/flightctl/pull/248
+
+#### Expected behavior
+
+When `fleet.spec.template` is updated the Fleet deployment changes, the fleet controller
+starts rolling out that update to devices. This update can take long (depending on the upgrade strategy)
+
+When a template is applied to a specific device via the Device API:
+* Floating tags for external references (os image, git refs, etc..) get converted to stable tags (hashes),
+  this gives flightctl the ability to check images being deployed on devices before the device could auto pick-up
+  a floating tag. (i.e. checks verifying the installation of the flightctl agent, and configuration)
+* When later in time a device has labels updated, the device controller will queue the device to be re-evaulated,
+  for example if a change of fleet owner happens because of the label changes.
+* An annotation is added to the device in the form of `fleet-controller/resourceVersion` to track the latest device
+  template applied to the device.
+
+> [!NOTE]
+> Devices also maintain a `metadata.resourceVersion` field, this is used to track changes related to re-applying
```

Comment:
see above; it's the API / database setting it, not the fleet controller.

Reply:
We will propose resourceVersion to all objects with metadata, just controlled at the storage level to make sure it updates every time. Then we can implement oportunistic concurrency access that we already identified we need with devices. (i.e. when somebody updates labels, but at the same time the fleet controller could be updating the spec...)